File: chapters/05_retry_and_recovery/v2_fixed_retry.ts
// TypeScript mirror for Chapter 05: Retry and Recovery (v2)
// Run with: npx ts-node v2_fixed_retry.ts

function runWithFixedRetry(
  simulatedErrors: string[],
  maxRetries: number = 3
): [string, string[]] {
  const logs: string[] = [];

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    logs.push(`attempt:${attempt}`);
    if (attempt < simulatedErrors.length) {
      logs.push(`error:${simulatedErrors[attempt]}`);
      continue;
    }
    logs.push("success");
    return ["success", logs];
  }

  logs.push("exhausted");
  return ["failed", logs];
}

function main(): void {
  console.log(runWithFixedRetry(["timeout", "timeout"], 2));
}

main();

File: chapters/05_retry_and_recovery/v2_fixed_retry.py
from __future__ import annotations


def run_with_fixed_retry(simulated_errors: list[str], max_retries: int = 3) -> tuple[str, list[str]]:
    logs: list[str] = []

    for attempt in range(max_retries + 1):
        logs.append(f"attempt:{attempt}")
        if attempt < len(simulated_errors):
            logs.append(f"error:{simulated_errors[attempt]}")
            continue
        logs.append("success")
        return "success", logs

    logs.append("exhausted")
    return "failed", logs


def main() -> None:
    print(run_with_fixed_retry(["timeout", "timeout"], max_retries=2))


if __name__ == "__main__":
    main()
